Add property machine.docker.local_node_host.external

l0rd opened the pull request:
### What does this PR do?
Introduce a new property, `machine.docker.local_node_host.external` (along with the environment variable `CHE_DOCKER_MACHINE_HOST_EXTERNAL`), that allow to configure the external URL browsers need to use to connect with containers that are not directly pingable. This happens for example with Docker for Mac.

```properties
machine.docker.local_node_host.external=localhost
```

When this property is set the browser will use it to connect to the containers. Note that the wsmaster won't use that value to connect to the containers.

If this property is not set Che behaves as before.

This PR introduce a change to the API: [Server class](https://github.com/eclipse/che/blob/master/core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/Server.java) has a couple of new attributes: `internalUrl` and `internalAddress`. On the other there no particular change is needed client side and all clients will benefit of this change.

```json
"servers": {
      "4401/tcp": {
                "url": "http://localhost:32906/wsagent/ext", 
                "internalUrl": "http://192.168.65.2:32906/wsagent/ext", 
                "address": "localhost:32906",
                "internalAddress": "192.168.65.2:32906",
                "path": "/wsagent/ext",
                "protocol": "http",
                "ref": "wsagent"
     }
}
```

This PR takes into account some feedbacks from PR #2004:
- There are no changes on client side
- Works on configurations where the wsagent and docker containers are on different hosts

I will not close PR #2004. I want to continue the work to integrate Traefik in the che-launcher there. But containous/traefik#444 should be fixed first.   

### What issues does this PR fix or reference?
#1644 

### Previous behavior
Running Che on Docker for Mac required some networking hacking

### New behavior
Remove the need for the networking hack when running Che on Docker for Mac and make running Che as simple as `docker run -v /var/run/docker.sock:/var/run/docker.sock codenvy/che-launcher start` on all platforms.

### PR type
- [x] Minor change = no change to existing features or docs
- [ ] Major change = changes existing features or docs

### Minor change checklist
- [ ] New API required?
- [x] API updated
- [x] Tests updated
- [x] Tests passed

TylerJewell commented:
Nice work!  Excited to have this get into master.  I reviewed the changes for the che-launcher and they all look good.

TylerJewell commented:
@garagatyi - this resolves a big inconvenience that users have with Docker for Mac. This eliminates a painful setup of the loopback alias on Macs. Your +1 here will be helpful.

garagatyi commented:
In my opinion changing API for that workaround is not a good way to go. It will add pain sooner or later.
I've just figured out that we can try to play with IPTables in Che container. 
For example we can try to redirect traffic inside of Che container from 32000-xxxxx to IP of xhyve machine. And let expose machines ports as localhost. In that case browser will be able to connect to machines on localhost and Che-server will use localhost also. But IPTables rules in che-server container will forward all traffic to another IP.
As I understand we can do that even after start of che-server, so we will be able to do that in che launcher and not change base che-server which works fine in all cases except `che in container on mac`.
Here is an example that I googled:
```iptables -t nat -A OUTPUT -i lo -p tcp --dport 32000:65000 -j DNAT --to <IP OF XHYVE>:32000-65000```
But I didn't try it and have no experience with IPTables, so not sure that I found correct (and complete) set of rules for out case.

l0rd commented:
@garagatyi can you please be more specific? Where do you want to run this `iptable` command: OSX, xyve VM, che-server container, agent container? What do you mean when you say "let expose machines ports as localhost"?

garagatyi commented:
Add iptables rules in che-server container.

Configure che-server in such a way that **localhost** is host in servers provided by Che API.
I think we just should not set CHE_DOCKER_MACHINE_HOST variable to achieve that.

So all traffic from che-server container to localhost with ports 32000-xxxxx will go through IPTables redirection to IP of xhyve where ports of all containers are available.

l0rd commented:
Ok thanks for clarifying @garagatyi. I understand now. It's a smart solution :-) I will test that tomorrow.

l0rd commented:
Bingo! I've struggled to apply the IP tables rules but I've finally found the following command that work if executed inside `che-sever` container!
```
apk update
apk install iptables
sysctl -w net.ipv4.conf.eth0.route_localnet=1
iptables -t nat -A OUTPUT -p tcp -d localhost -o lo --dport 32000:65000 -j DNAT --to ${XHYVE_VM_IP}:32000-65000
iptables -t nat -A POSTROUTING -o eth0 -m addrtype --src-type LOCAL --dst-type UNICAST -j MASQUERADE
```

l0rd commented:
I've created a new PR to address the Docker for Mac problem: eclipse/che-dockerfiles#12

@garagatyi I will let you choose which one of these 2 PR is more suitable but let me explain why I believe we should keep this one and close eclipse/che-dockerfiles#12. Apart [some concerns](https://github.com/eclipse/che-dockerfiles/pull/12#issuecomment-246398648) with the iptables solution I also think that changing the Server class, adding an internal url attribute, is not just a patch to solve the Docker for Mac issue but a solution for a recurring problem.

![image](https://cloud.githubusercontent.com/assets/606959/18443241/4ce96fe4-7914-11e6-862b-4106153dfabc.png)

In the current model there is only one address to reach the wsagent server. That same address is used by the browser and the wsmaster. In the diagram I've called this address "Post Office". Depending on the platform the "Post Office" can be `docker0`, the IP of the docker host or just `localhost`. However in most scenarios wsmaster and wsagent are close and there is a shorter path to reach the wsagent (the red path in the diagram). Most important this red path avoid some scary mountains that can be tricky to cross (firewalls, routers, IPtables etc...). That's the case with Docker for Mac but that's also the case with Ubuntu (#2030) or Kubernetes or Openshift etc...This PR is to add the address of this red path to the attributes of the Server class.

TylerJewell commented:
I tend to agree with Mario that having the internal sokution is more optimal for the longer term.  And that we can continue to layer in the iptables solution where necessary.

Great graphic :)

l0rd commented:
I've updated the PR. I've added a new ServerProperties class where I moved servers `path`, `internalUrl` and `internalAddress` fields. That's the new Server API:

```json
"servers": {
    "4401/tcp": {
        "ref": "wsagent",
        "protocol": "http",
        "address": "localhost:32929",
        "url": "http://localhost:32929/wsagent/ext",
        "port": "32929",
        "properties": {
            "internalUrl": "http://192.168.65.2:32929/wsagent/ext",
            "internalAddress": "192.168.65.2:32929",
            "path": "/wsagent/ext"
        }
    }
}
```
@garagatyi  please review

benoitf commented:
I like it also
 I would say that timeframe for updating json format as we're in process for 5.x release and still in milestone,is the right time. After it would be too late

TylerJewell commented:
I am ok with this approach.  I want to target this for M2 - solving this problem simplifies a lot of scenarios for us with Mac.

garagatyi commented on core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/ServerProperties.java:
```diff
@@ -0,0 +1,41 @@
+     * Internal address of the server in form <b>host:port</b>
+     * Used by wsmaster to communicate with the server
+     */
+    String getInternalAddress();
```
Looks like all fields in that object are nullable.

garagatyi commented on core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/ServerProperties.java:
```diff
@@ -0,0 +1,41 @@
+    String getPath();
+
+    /**
+     * Internal address of the server in form <b>host:port</b>
```
Line break by entering 'enter' key doesn't work in javadocs. Please add dot and `<p/>` or `<br/>`

garagatyi commented on core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/Server.java:
```diff
@@ -25,25 +25,35 @@
 
     /**
-     * Address of the server in form <b>host:port</b>
+     * External address of the server in form <b>host:port</b>
```
Line break by entering 'enter' key doesn't work in javadocs. Please add dot and `<p/>` or `<br/>`

garagatyi commented on ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/machine/DevMachineServerProperties.java:
```diff
@@ -0,0 +1,47 @@
+import org.eclipse.che.api.core.model.machine.ServerProperties;
+
+/**
+ * Describe development machine server instance.
```
Line break by entering 'enter' key doesn't work in javadocs. Please add `<p/>` or `<br/>`

garagatyi commented on plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceRuntimeInfo.java:
```diff
@@ -82,17 +83,20 @@
     @Inject
     public DockerInstanceRuntimeInfo(@Assisted ContainerInfo containerInfo,
-                                     @Assisted String containerHost,
+                                     @Assisted String containerExternalHostname,
```
Not named assisted injection doesn't work when several arguments have the same class

garagatyi replied:
Also please check what will happen if someone provide null value as one of these parameters. I believe nullable annotation should be declared somewhere.

l0rd replied:
I've changed to named assisted injection and checked if `containerExternalHostname` is null and set it to `containerHost` in that case.

garagatyi commented on core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/Server.java:
```diff
@@ -25,25 +25,35 @@
     String getAddress();
 
     /**
+     * Port of the server, e.g. 8080
```
I'm not sure should this port contain `8080` or `8080/tcp`.

garagatyi replied:
And it is external port. Please explain that in javadocs.

l0rd replied:
It's `8080`, the protocol is not there:
```json
"port": "32929"
```

garagatyi commented on ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/machine/DevMachineServer.java:
```diff
@@ -53,12 +56,13 @@ public String getProtocol() {
+        public String getPort() { return port; }
+
+        @Override
+        public ServerProperties getProperties() { return properties; };
```
Please add equals/hashcode/toString methods

l0rd replied:
Done

garagatyi commented on ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/machine/DevMachineServerProperties.java:
```diff
@@ -0,0 +1,47 @@
+    public String getInternalUrl() {
+        return internalUrl;
+    }
+}
```
Please add equals/hashcode/toString methods

l0rd replied:
Done

garagatyi commented on ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/machine/DevMachineServerProperties.java:
```diff
@@ -0,0 +1,47 @@
+    private final String internalAddress;
+    private final String internalUrl;
+
+    public DevMachineServerProperties(ServerProperties dto) {
```
Argument doesn't have to be DTO object. Please rename parameter to not confuse developers.

l0rd replied:
Done

garagatyi commented on ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/machine/DevMachineServerProperties.java:
```diff
@@ -0,0 +1,47 @@
+
+/**
+ * Describe development machine server instance.
+ * {@link ServerProperties}
```
Curly brackets are used when link is inlined into description. When it is used as a separate tag it has the same syntax as `@author`

l0rd replied:
Done

garagatyi commented on plugins/plugin-machine/che-plugin-machine-ext-client/src/main/java/org/eclipse/che/ide/extension/machine/client/perspective/widgets/machine/appliance/server/ServerProperties.java:
```diff
@@ -0,0 +1,46 @@
+    @Override
+    public String getInternalUrl() {
+        return descriptor.getInternalUrl();
+    }
```
Please add equals/hashcode/toString methods

garagatyi replied:
This class still needs equals/hashcode/toString methods

l0rd replied:
Class https://github.com/eclipse/che/blob/master/plugins/plugin-machine/che-plugin-machine-ext-client/src/main/java/org/eclipse/che/ide/extension/machine/client/perspective/widgets/machine/appliance/server/Server.java had no equals method. Should I add equals/hashcode/toString to existing classes too?

garagatyi replied:
At least add it in classes added by you.

l0rd replied:
I've added for both.

garagatyi commented on core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/Server.java:
```diff
@@ -25,25 +25,35 @@
+     * Can be set using property machine.docker.local_node_host.external or
+     * environment variable CHE_DOCKER_MACHINE_HOST_EXTERNAL
      */
     String getAddress();
```
Now this method returns both host and port of the server. Are you going to change the behavior in this PR?

l0rd replied:
No I'm not:
```json
"address": "localhost:32929"
```
Is that ok?

garagatyi commented:
Looks like I'm not fully understand new server object. I'll take a look again tomorrow and ask you if something is not clear.

l0rd commented:
@garagatyi I've commited the fixes you have suggested:
- Javadoc improved
- Removed `port` field in `Server` class
- Annotated all fields in `ServerProperties` as nullable
- Overwritten methods `equals`, `hashcode` and  `toString` for classes `DevMachineServer` and `DevMachineServerProperties`
- Fixed unnamed assisted injections
- Added some unit tests
- Restored the license header (I'll submit a distinct PR for that)
- Environment variables have now higher priority then properties

garagatyi commented on core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/Server.java:
```diff
@@ -36,14 +42,15 @@
 
     /**
-     * Path to access the server.
+     * Url of the server, e.g.&nbsp;http://localhost:8080
```
Please replace `&nbsp;` with a regular backspace.

l0rd replied:
I think you mean "regular space". If I add a regular space and I try to generate the Javadoc I have the following result:

![image](https://cloud.githubusercontent.com/assets/606959/18626032/b4237b6c-7e52-11e6-8797-72cb966a3414.png)

If I use `&nbsp;` instead the sentence is not broken:

![image](https://cloud.githubusercontent.com/assets/606959/18625990/70ca2776-7e52-11e6-962c-376815b3153b.png)

garagatyi replied:
OK, interesting

garagatyi commented on plugins/plugin-machine/che-plugin-machine-ext-client/src/test/java/org/eclipse/che/ide/extension/machine/client/perspective/widgets/machine/appliance/server/ServerTest.java:
```diff
@@ -81,4 +80,37 @@ public void refShouldBeReturned() {
+        verify(descriptor2).getInternalUrl();
+    }
+
 }
```
Please add trailing new line at the end of file

l0rd replied:
Fixed

garagatyi commented on wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/DtoConverter.java:
```diff
@@ -10,24 +10,8 @@
  *******************************************************************************/
 package org.eclipse.che.api.machine.server;
 
-import org.eclipse.che.api.core.model.machine.MachineLimits;
```
Wildcard imports are forbidden in Che

l0rd replied:
Sorry for forgetting that. I've fixed it.

garagatyi commented on wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/WorkspaceServiceTest.java:
```diff
@@ -24,14 +24,7 @@
 import org.eclipse.che.api.core.rest.shared.dto.ServiceError;
 import org.eclipse.che.api.environment.server.MachineProcessManager;
 import org.eclipse.che.api.environment.server.MachineServiceLinksInjector;
-import org.eclipse.che.api.machine.server.model.impl.CommandImpl;
```
Please do not use wildcard imports

l0rd replied:
Fixed

garagatyi commented on plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstance.java:
```diff
@@ -145,6 +145,7 @@ public MachineRuntimeInfoImpl getRuntime() {
             try {
                 final ContainerInfo containerInfo = docker.inspectContainer(container);
                 machineRuntime = new MachineRuntimeInfoImpl(dockerMachineFactory.createMetadata(containerInfo,
+                                                                                                null,
```
For me it looks like this argument is always null. Am I miss something?

l0rd replied:
Feel free to add more details to your questions if that was not what you meant. I think you are wondering how `Server.url` is ever set if this parameter is always `null`. It should be set using property `machine.docker.local_node_host.external` but it looks as it's always null.

Property `machine.docker.local_node_host.external` is injected in `LocalDockerModule` objects that are binded to `DockerInstanceRuntimeInfo` in `LocalDockerModule`. 

Therefore method `DockerInstance.getRuntime()` is not used by the API to get the machine runtime. That explains why `Server.url` is correctly returned by the API.

garagatyi replied:
Let me explain how I understand what is happening with these changes. Please correct me if I'm mistaken.
DockerInstance always pass null as external host. DockerInstanceRuntimeInfo doesn't change that. LocalDockerInstanceRuntimeInfo uses assisted injection, so this null is passed here too. 
So I don't see how another than null value can be accepted by DockerInstanceRuntimeInfo.

l0rd replied:
As discussed on Slack I've cleaned-up `LocalDockerInstanceRuntimeInfo` class and made it easier to read.

l0rd commented on wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/WorkspaceServiceTest.java:
```diff
@@ -24,14 +24,7 @@
-import org.eclipse.che.api.machine.server.model.impl.MachineSourceImpl;
-import org.eclipse.che.api.machine.server.model.impl.ServerImpl;
-import org.eclipse.che.api.machine.server.model.impl.SnapshotImpl;
+import org.eclipse.che.api.machine.server.model.impl.*;
```
I think I alrady fixed that: https://github.com/l0rd/che/blob/a35ccae37ab7fdecf958beb02acbf0e8177c8a3b/wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/WorkspaceServiceTest.java

You were probably looking at the outdated version?

benoitf replied:
yep it was pending request (I forgot to validate it)

l0rd commented on wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/model/impl/ServerImpl.java:
```diff
@@ -22,22 +23,24 @@
 
     public ServerImpl(Server server) {
-        this(server.getRef(), server.getProtocol(), server.getAddress(), server.getPath(), server.getUrl());
+        this(server.getRef(), server.getProtocol(), server.getAddress(), server.getUrl(), server.getPort(), (ServerPropertiesImpl) server.getProperties());
```
Fixed

l0rd commented on plugins/plugin-docker/che-plugin-docker-machine/src/test/java/org/eclipse/che/plugin/docker/machine/DockerInstanceRuntimeInfoTest.java:
```diff
@@ -507,30 +574,43 @@ public void shouldAddCommonAndDevSystemServersConfigToDevMachine() throws Except
+                                                       "32103",
+                                                       new ServerPropertiesImpl("some/path4",
+                                                           CONTAINER_HOST + ":32103",
+                                                           null)));
```
Added a couple of tests

l0rd commented on plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/local/LocalDockerInstanceRuntimeInfo.java:
```diff
@@ -24,32 +24,52 @@
-              dockerNodeHost != null ? dockerNodeHost : (System.getenv(CHE_DOCKER_MACHINE_HOST) != null ?
-                                                         System.getenv(CHE_DOCKER_MACHINE_HOST) :
-                                                         containerHost),
+              dockerNodeExternalHostname != null ? dockerNodeInternalHostname : (
```
Fixed. Env variables are now evaluated first.

l0rd commented on plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerInstanceRuntimeInfo.java:
```diff
@@ -299,9 +315,10 @@ public String projectsRoot() {
                                                      null,
-                                                     null));
+                                                     externalPort,
+                                                     new ServerPropertiesImpl(null, internalHostname + ":" + externalPort, null)));
```
After fixup `externalHostname`, `internalHostname` and `externalPorts` can never be `null`.

garagatyi commented:
@musienko-maxim Please do a QA cycle for this PR

garagatyi commented:
ci-build

garagatyi commented:
QA team has approved this PR. We will merge it when M3 is released
